vaadin-text-field-flow-parent/vaadin-text-field-flow-integration-tests/src/test/java/com/vaadin/flow/component/textfield/tests/validation/NumberFieldValidationBasicIT.java

```diff
@@ -0,0 +1,185 @@
+
+        detachAndReattachField();
+
+        assertWebComponentCanNotModifyInvalidState();
+    }
+
+    @Test
+    public void triggerInputBlur_assertValidity() {
+        testField.sendKeys(Keys.TAB);
+        assertServerValid();
+        assertClientValid();
+    }
+
+    @Test
+    public void required_triggerInputBlur_assertValidity() {
+        $("button").id(REQUIRED_BUTTON).click();
+
+        testField.sendKeys(Keys.TAB);
+        assertServerInvalid();
+        assertClientInvalid();
+    }
+
+    @Test
+    public void required_changeInputValue_assertValidity() {
+        $("button").id(REQUIRED_BUTTON).click();
+
+        testField.setValue("1234");
+        assertServerValid();
+        assertClientValid();
+
+        testField.setValue("");
+        assertServerInvalid();
+        assertClientInvalid();
+    }
+
+    @Test
+    public void min_triggerInputBlur_assertValidity() {
+        $("input").id(MIN_INPUT).sendKeys("2", Keys.ENTER);
+
+        testField.sendKeys(Keys.TAB);
+        assertServerValid();
+        assertClientValid();
+    }
+
+    @Test
+    public void min_changeInputValue_assertValidity() {
+        $("input").id(MIN_INPUT).sendKeys("2", Keys.ENTER);
+
+        testField.setValue("1.8");
+        assertClientInvalid();
+        assertServerInvalid();
+
+        testField.setValue("2");
+        assertClientValid();
+        assertServerValid();
+
+        testField.setValue("2.2");
+        assertClientValid();
+        assertServerValid();
+    }
+
+    @Test
+    public void max_triggerInputBlur_assertValidity() {
+        $("input").id(MAX_INPUT).sendKeys("2", Keys.ENTER);
+
+        testField.sendKeys(Keys.TAB);
+        assertServerValid();
+        assertClientValid();
+    }
+
+    @Test
+    public void max_changeInputValue_assertValidity() {
+        $("input").id(MAX_INPUT).sendKeys("2", Keys.ENTER);
+
+        testField.setValue("2.2");
+        assertClientInvalid();
+        assertServerInvalid();
+
+        testField.setValue("2");
+        assertClientValid();
+        assertServerValid();
+
+        testField.setValue("1.8");
+        assertClientValid();
+        assertServerValid();
+    }
+
+    @Test
+    public void step_triggerInputBlur_assertValidity() {
+        $("input").id(STEP_INPUT).sendKeys("2", Keys.ENTER);
+
+        testField.sendKeys(Keys.TAB);
+        assertServerValid();
+        assertClientValid();
+    }
+
+    @Test
+    public void step_changeInputValue_assertValidity() {
+        $("input").id(STEP_INPUT).sendKeys("1.5", Keys.ENTER);
+
+        testField.setValue("1");
+        assertClientInvalid();
+        assertServerInvalid();
+
+        testField.setValue("1.5");
+        assertClientValid();
+        assertServerValid();
+
+        testField.setValue("2");
+        assertClientInvalid();
+        assertServerInvalid();
+    }
+
+    @Test
+    public void badInput_changeInputValue_assertValidity() {
+        testField.sendKeys("--2", Keys.TAB);
```

Comment:
**note:** Unlike other tests, I had to use <kbd>Tab</kbd> here to commit the value. <kbd>Enter</kbd> didn't work due to the issue:

- https://github.com/vaadin/web-components/issues/5113